apps/ae_socket_connector/lib/client_runner.ex

```diff
@@ -103,35 +107,85 @@ defmodule ClientRunner do
     end
   end
 
-  def process_sign_request(message, to_sign, state) do
-    try do
-      case elem(message, 0) do
-        # TODO how do we descide if we should sign?
-        :sign_approve ->
-          signed = SessionHolder.sign_message(state.pid_session_holder, to_sign)
-          fun = fn pid -> SocketConnector.send_signed_message(pid, elem(message, 2), signed) end
-          SessionHolder.run_action(state.pid_session_holder, fun)
+  def process_sign_request(message, to_sign, pid_session_holder, mode \\ %{sign: {:default}})
 
-        _ ->
-          :ok
-      end
-    rescue
-      _ignore -> :ok
+  def process_sign_request({}, _to_sign, _pid_session_holder, _mode) do
+    Logger.debug("Empty request")
+  end
+
+  def process_sign_request(message, to_sign, pid_session_holder, %{sign: sign_info}) do
+    case elem(message, 0) do
+      :sign_approve ->
+        case sign_info do
+          {:default} ->
+            signed = SessionHolder.sign_message(pid_session_holder, to_sign)
+            fun = fn pid -> SocketConnector.send_signed_message(pid, elem(message, 2), signed) end
+            SessionHolder.run_action(pid_session_holder, fun)
+
+          {:backchannel, pid_other_session_holder} ->
+            signed = SessionHolder.sign_message(pid_session_holder, to_sign)
+            signed2 = SessionHolder.sign_message(pid_other_session_holder, signed)
+            fun = fn pid -> SocketConnector.send_signed_message(pid, elem(message, 2), signed2) end
+            SessionHolder.run_action(pid_session_holder, fun)
+
+          {:check_poi} ->
+            fun = fn pid, from -> SocketConnector.get_poi(pid, from) end
+            poi = SessionHolder.run_action_sync(pid_session_holder, fun)
+            case SessionHolder.verify_poi(pid_session_holder, to_sign, poi) do
+              :ok ->
+                signed = SessionHolder.sign_message(pid_session_holder, to_sign)
+                fun = fn pid -> SocketConnector.send_signed_message(pid, elem(message, 2), signed) end
+                SessionHolder.run_action(pid_session_holder, fun)
+              :unsecure ->
+                Logger.warn("POI missmatch, refuse signing")
+            end
+
+          {:abort, abort_code} ->
+            method = elem(message, 2)
+            fun = fn pid -> SocketConnector.abort(pid, method, abort_code, "some message") end
+            SessionHolder.run_action(pid_session_holder, fun)
+
+          _ ->
+            Logger.debug("Don't sign")
+        end
+
+      _ ->
+        :ok
     end
   end
 
-  # {:responder, :channels_info, 0, :transient, "channel_open"}
-  # def handle_cast({:match_jobs, message}, state)
-  def handle_cast({:match_jobs, received_message, to_sign}, state) do
-    process_sign_request(received_message, to_sign, state)
+  def process_sign_request(message, to_sign, pid_session_holder, _not_sign_request) do
+    process_sign_request(message, to_sign, pid_session_holder)
+  end
+
```

Comment:
General code layout: You mix public functions with gen_server callback implementations. I'd rather see some organization to keep these as groups separated with a comment before the groups.

Reply:
Agree will fix this later on.